src/shared/telemetry/defaultTelemetryPublisher.ts

```diff
@@ -0,0 +1,113 @@
+
+import { CognitoIdentity, CognitoIdentityCredentials } from 'aws-sdk'
+import { DefaultTelemetryClient } from './defaultTelemetryClient'
+import { TelemetryClient } from './telemetryClient'
+import { TelemetryEvent, TelemetryEventArray } from './telemetryEvent'
+import { TelemetryPublisher } from './telemetryPublisher'
+
+export interface IdentityPublisherTuple {
+    cognitoIdentityId: string,
+    publisher: TelemetryPublisher
+}
+
+export class DefaultTelemetryPublisher implements TelemetryPublisher {
+    private static readonly DEFAULT_MAX_BATCH_SIZE = 20
+
+    private readonly eventQueue: TelemetryEventArray
+
+    public constructor(
+        private readonly clientId: string,
+        private readonly region: string,
+        private readonly credentials: AWS.Credentials,
+        private telemetryClient?: TelemetryClient
+    ) {
+        this.eventQueue = new TelemetryEventArray()
+    }
+
+    public enqueue(events: TelemetryEvent[]): void {
+        this.eventQueue.push(...events)
+    }
+
+    public getQueue(): ReadonlyArray<TelemetryEvent> {
+        return this.eventQueue
+    }
+
+    public async flush(): Promise<void> {
+        if (this.telemetryClient === undefined) {
+            await this.init()
+        }
+
+        while (this.eventQueue.length !== 0) {
+            const batch = this.eventQueue
+                .splice(0, DefaultTelemetryPublisher.DEFAULT_MAX_BATCH_SIZE) as TelemetryEventArray
+
+            if (this.telemetryClient === undefined) {
+                return
+            }
+
+            const failedBatch = await this.telemetryClient.postMetrics(batch)
+            if (failedBatch !== undefined) {
+                this.enqueue(failedBatch)
+
+                // retry next time
+                return
+            }
+        }
+    }
+
+    public async init(): Promise<void> {
+        this.telemetryClient = await DefaultTelemetryClient.createDefaultClient(
+            this.clientId,
+            this.region,
+            this.credentials
+        )
+    }
+
+    public static async fromDefaultIdentityPool(clientId: string): Promise<IdentityPublisherTuple> {
+        return this.fromIdentityPool(clientId, DefaultTelemetryClient.DEFAULT_IDENTITY_POOL)
+    }
+
+    /**
+     * Create a telemetry publisher from the given clientId and identityPool
+     * @return A tuple containing the new identityId and the telemetry publisher
+     */
+    public static async fromIdentityPool(clientId: string, identityPool: string)
+        : Promise<IdentityPublisherTuple> {
+        const region = identityPool.split(':')[0]
+        try {
+            const res = await new CognitoIdentity({
+                region: region
+            }).getId({
+                IdentityPoolId: identityPool
+            }).promise()
+            const err = res.$response.error
+            if (err) {
+                return Promise.reject('SDK deserialization error')
+            }
+            const identityId = res.IdentityId!!
```

Comment:
Why is there a `!!` here?

Reply:
`IdentityId` can be `undefined` if there was any error in the Cognito.getId request

Reply:
1. Non-undefined assertions in TypeScript just use one character: `res.IdentityId!`
2. Rather than non-undefined asserting, can you test `res.IdentityId` and throw if it's falsy?

Reply:
:+1: will throw if falsy